experimental/transliteration/src/transliterator/mod.rs

```diff
@@ -468,55 +458,38 @@ impl<'a> RuleBasedTransliterator<'a> {
         // assumes the cursor is at the right position.
 
-        debug_assert_eq!(self.id_group_list.len(), self.rule_group_list.len());
-        debug_assert!(rep.allowed_range().contains(&rep.cursor()));
-
         rep.for_each_run(&self.filter, |run| {
             // eprintln!("got RBT filtered_run: {run:?}");
-            self.transliterate_run(run, env);
-            // eprintln!("finished RBT filtered_run transliteration: {run:?}")
-        });
-    }
+            for (id_group, rule_group) in self.id_group_list.iter().zip(self.rule_group_list.iter())
+            {
+                // first handle id_group
+                for single_id in id_group.iter() {
+                    let id = SimpleId::zero_from(single_id);
+                    id.transliterate(run.child(), env);
+                }
 
-    /// Transliteration of a single run, i.e., without needing to look at the filter.
-    fn transliterate_run(&self, rep: &mut Replaceable, env: &Env) {
-        // assumes the cursor is at the right position.
-        // note that there is no such thing as an empty run.
-        debug_assert!(
-            rep.allowed_range().contains(&rep.cursor()),
-            "cursor {} is out of bounds for replaceable {rep:?}",
-            rep.cursor()
-        );
-
-        for (id_group, rule_group) in self.id_group_list.iter().zip(self.rule_group_list.iter()) {
-            // first handle id_group
-            for single_id in id_group.iter() {
-                let id = SimpleId::zero_from(single_id);
-                id.transliterate(rep.child(), env);
+                // then handle rule_group
+                let rule_group = RuleGroup::from(rule_group);
+                rule_group.transliterate(run.child(), &self.variable_table, env);
             }
-
-            // then handle rule_group
-            let rule_group = RuleGroup::from(rule_group);
-            rule_group.transliterate(rep.child(), &self.variable_table, env);
-        }
+            // eprintln!("finished RBT filtered_run transliteration: {run:?}")
+        });
     }
 }
 
 impl<'a> SimpleId<'a> {
     fn transliterate(&self, mut rep: Replaceable, env: &Env) {
         // eprintln!("transliterating SimpleId: {self:?}");
-        rep.for_each_run(&self.filter, |run| self.transliterate_run(run, env))
-    }
-
-    fn transliterate_run(&self, rep: &mut Replaceable, env: &Env) {
-        // eprintln!("transliterating SimpleId run: {rep:?}");
-        match env.get(self.id.as_ref()) {
-            None => {
-                debug_assert!(false, "missing transliterator {}", &self.id);
-                // GIGO behavior, missing recursive transliterator is a noop
+        rep.for_each_run(&self.filter, |run| {
+            // eprintln!("transliterating SimpleId run: {rep:?}");
+            match env.get(self.id.as_ref()) {
+                None => {
+                    debug_assert!(false, "missing transliterator {}", &self.id);
```

Comment:
Is this actually possible? You're already erroring in `load_dependencies_recursive`, so all dependencies will be present at this point.

Reply:
Good point!

Reply:
Actually, no. We load the _dependencies_, but corrupt data may not have an in-sync `dependencies` list with its actual dependencies.

The only place dependencies exist are in `RBT.id_group_list` and `VarTable.function_calls`. If we wanted to check that validity at construction time, we could iterate over those, at which point the explicit `dependencies` list is bloat and we can remove it to save some space. Thoughts? I'd do that in a follow-up, not this PR.

Reply:
You're right, in that case GIGO is the correct approach here. The optimisation sounds good for a follow-up.